src/Networking/NexusMods.Networking.Steam/Session.cs

```diff
@@ -0,0 +1,250 @@
+        _steamUser = _steamClient.GetHandler<SteamUser>()!;
+        _steamApps = _steamClient.GetHandler<SteamApps>()!;
+        _steamContent = _steamClient.GetHandler<SteamContent>()!;
+        _cdnClient = new Client(_steamClient);
+        _cdnPool = new CDNPool(this);
+        
+        // Some parts of this interface use callbacks instead of more natural async methods. So we need to register
+        // those callbacks here.
+        _callbacks = new CallbackManager(_steamClient);
+        _callbacks.Subscribe(WrapAsync<SteamClient.ConnectedCallback>(ConnectedCallback));
+        _callbacks.Subscribe(WrapAsync<SteamClient.DisconnectedCallback>(DisconnectedCallback));
+        _callbacks.Subscribe(WrapAsync<SteamUser.LoggedOnCallback>(LoggedOnCallback));
+        _callbacks.Subscribe(WrapAsync<SteamApps.LicenseListCallback>(LicenseListCallback));
+    }
+
+    /// <summary>
+    /// The Steam Content module
+    /// </summary>
+    internal SteamContent Content => _steamContent;
+    
+    /// <summary>
+    /// The CDN client, used for downloading game data.
+    /// </summary>
+    internal Client CDNClient => _cdnClient;
+    
+    /// <summary>
+    /// The CDN pool, used for downloading game data.
+    /// </summary>
+    internal CDNPool CDNPool => _cdnPool;
+
+    private Task LicenseListCallback(SteamApps.LicenseListCallback arg)
+    {
+        return Task.CompletedTask;
+    }
+    
+    private async Task LoggedOnCallback(SteamUser.LoggedOnCallback callback)
+    {
+        _isLoggedOn = true;
+        _logger.LogInformation("Logged on to Steam network.");
+    }
+
+    private async Task DisconnectedCallback(SteamClient.DisconnectedCallback callback)
+    {
+        _logger.LogInformation("Disconnected from Steam network.");
+    }
+    
+    private async Task ConnectedCallback(SteamClient.ConnectedCallback callback)
+    {
+        _isConnected = true;
+        var (success, data) = await _authStorage.TryLoad();
+        if (success)
+        {
+            _logger.LogInformation("Using saved auth data to log in.");
+            var authData = AuthData.Load(data);
+            
+            _steamUser.LogOn(new SteamUser.LogOnDetails
+                {
+                    Username = authData.Username,
+                    AccessToken = authData.RefreshToken,
+                }
+            );
+        }
+        else
+        {
+            _logger.LogInformation("No saved auth data, logging in via QR code.");
+            var authSession = await _steamClient.Authentication.BeginAuthSessionViaQRAsync(new AuthSessionDetails());
+
+            authSession.ChallengeURLChanged = () => { _handler.ShowQRCode(new Uri(authSession.ChallengeURL, UriKind.Absolute), CancellationToken.None); };
+
+            _handler.ShowQRCode(new Uri(authSession.ChallengeURL, UriKind.Absolute), CancellationToken.None);
+            var pollResponse = await authSession.PollingWaitForResultAsync();
```

Comment:
This accepts a cancellation token, we should have a timeout here:


```suggestion
            using var cts = new CancellationTokenSource();
            cts.CancelAfter(delay: TimeSpan.FromMinutes(5));

            var pollResponse = await authSession.PollingWaitForResultAsync(cancellationToken: cts.Token);
```